### ISSUE TITLE
Improve cannot assign warning on table literals in regards to unions

### ISSUE THREAD

**Issue by PennyJim** [EmmyLuaLs/emmylua-analyzer-rust#1168]
Currently, if you are assigning to a nilable field and some sub-type is not assignable, it just is all masked into the top level "Cannot assign"
This makes it hard to catch that one of the values is slightly wrong when you don't realize it. It's easy to just assume EmmyLua is confused about something and making a bogus warning, especially on more complex things.

```lua
---@class Icon
---@field icon string
---@class IconList
---@field one Icon

---@class A
---@field icon? string
---@field icons? Icon[]
---@field icon_list? IconList

---@param a A
function test(a) end

---@type A
local tmp

tmp.icon_list = { -- Cannot assign `{ one = table }` to `IconList?`.
  one = {icon = tmp.icon},
}

tmp.icons = { -- Cannot assign `(table)` to `Icon[]?`.
  {icon = tmp.icon}
}

test{
  icons = { -- Cannot assign `(table)` to `Icon[]?`.
    {icon = tmp.icon}
  },
  icon_list = { -- Cannot assign `{one = table}` to `IconList?`.
    one = {icon = tmp.icon}
  }
}
```

When the field being assigned to is *not* a union with nil, it has a better warnings. The arrays entries are still not great, but I feel like they're a symptom of arrays seemingly being `table<int,A?>` and therefore inheriting the issue with a union to nil.
```lua
---@class B
---@field icon? string
---@field icons Icon[]
---@field icon_list IconList

---@param b B
local function test(b) end

---@type B
local tmp


tmp.icon_list = {
  one = {icon = tmp.icon}, -- Cannot assign `string?` to `string`.
}

tmp.icons = { -- Cannot assign `(table)` to `Icon[]?`.
  {icon = tmp.icon} -- Cannot assign `{ icon = string? }` to `Icon?`.
}

test{
  icons = {
    {icon = tmp.icon} -- Cannot assign `{ icon = string? }` to `Icon?`.
  },
  icon_list = {
    one = {icon = tmp.icon} -- Cannot assign `string?` to `string`.
  }
}
```

This has specifically caused us lots of strife with multi-form named parameter style functions:
```lua
---@class C
---@field type "one"
---@class D
---@field type "two"

---@param cd C|D
local function cd(cd) end
---@param c C
local funcion c(c) end

cd{ -- expected `(C|D)` but found `{ type = "test" }`.
  type = "test"
}

c{
  type = "test" -- Cannot assign `string` to `string`.
}
```
This also showcases the secondary issue of assigning a string literal to a different string literal.

These problems are compounded when the objects in question are quite large and therefore difficult to find the problem, like the ones in the Factorio api.
Key example being the `param-type-mismatch` warnings on our [`LuaGuiElement.add{}`](https://lua-api.factorio.com/latest/classes/LuaGuiElement.html#add).

**Comment by PennyJim** [comment]
While looking through the tangzx extension issues to prepare to report a bug, I found this related report covering functions.

https://github.com/EmmyLua/VSCode-EmmyLua/issues/308

**Comment by xuhuanzy** [comment]
I refactored the type-checking system to address this issue, so fixing it will take longer.
